packages/cw0/src/parse_reply.rs

```diff
@@ -0,0 +1,214 @@
+use thiserror::Error;
+
+use cosmwasm_std::{Binary, Reply};
+
+// Protobuf wire types (https://developers.google.com/protocol-buffers/docs/encoding)
+const WIRE_TYPE_LENGTH_DELIMITED: u8 = 2;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MsgInstantiateContractResponse {
+    pub contract_address: String,
+    pub data: Option<Binary>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct MsgExecuteContractResponse {
+    pub data: Option<Binary>,
+}
+
+/// Helper function to parse length-prefixed protobuf fields.
+/// The remaining of the data is kept in the data parameter.
+fn parse_protobuf_length_prefixed(data: &mut Vec<u8>) -> Result<Vec<u8>, ParseReplyError> {
+    if data.is_empty() {
+        return Ok(vec![]);
+    };
+    let mut rest_1 = data.split_off(1);
+    let wire_type = data[0] & 0b11;
+    let field = data[0] >> 3;
+
+    if wire_type != WIRE_TYPE_LENGTH_DELIMITED {
+        return Err(ParseReplyError::ParseFailure(format!(
+            "failed to decode Protobuf message: field #{}: invalid wire type {}",
+            field, wire_type
+        )));
+    }
+    if rest_1.is_empty() {
+        return Err(ParseReplyError::ParseFailure(format!(
+            "failed to decode Protobuf message: field #{}: message too short",
+            field
+        )));
+    }
+    let mut rest_2 = rest_1.split_off(1);
+    let len = rest_1[0] as usize;
```

Comment:
This is valid for sizes <= 127 bytes.

You can either:

* add a caveat for that as rust comment and issue to follow up (but return error if this is multi-byte - if `rest1[0] & 0x80 != 0`)
* support this now as defined here: https://developers.google.com/protocol-buffers/docs/encoding#varints

Reply:
By the way, I thought this is was an inefficient / inconvenient format for length encoding, and this seems to confirm it: https://news.ycombinator.com/item?id=11656973

Reply:
I suggest we cap it at two / three bytes, and error otherwise.